rust/rust-sysdb/src/spanner.rs

```diff
@@ -25,27 +35,293 @@ impl ChromaError for SpannerError {
+            )));
+        }
+
+        // Use a read-write transaction to atomically check and insert
+        self.client
+            .read_write_transaction(|tx| {
+                let db_id = req.id.clone();
+                let db_name = req.name.clone();
+                let tenant_id = req.tenant.clone();
+                Box::pin(async move {
+                    // Check if database already exists
+                    let mut check_stmt = Statement::new(
+                        "SELECT ID FROM databases WHERE ID = @id AND IS_DELETED = FALSE",
+                    );
+                    check_stmt.add_param("id", &db_id);
+
+                    let mut iter = tx.query(check_stmt).await?;
+
+                    // If database doesn't exist, insert it
+                    if iter.next().await?.is_none() {
+                        let mut insert_stmt = Statement::new(
+                            "INSERT INTO databases (ID, NAME, TENANT_ID, IS_DELETED, CREATED_AT, UPDATED_AT) VALUES (@id, @name, @tenant_id, @is_deleted, PENDING_COMMIT_TIMESTAMP(), PENDING_COMMIT_TIMESTAMP())",
+                        );
+                        insert_stmt.add_param("id", &db_id);
+                        insert_stmt.add_param("name", &db_name);
+                        insert_stmt.add_param("tenant_id", &tenant_id);
+                        insert_stmt.add_param("is_deleted", &false);
+
+                        tx.update(insert_stmt).await?;
+                        tracing::info!("Created database: {} for tenant: {}", db_name, tenant_id);
+                    } else {
+                        tracing::debug!("Database already exists, skipping insert: {}", db_id);
+                    }
+
+                    Ok(())
+                })
+            })
+            .await
+            .map_err(|e: SpannerClientError| SysDbError::Spanner(e.to_string()))?;
+
+        Ok(CreateDatabaseResponse {})
+    }
+
+    /// Get a database by name and tenant.
+    ///
+    /// Returns None if the database does not exist or is marked as deleted.
+    pub async fn get_database(
+        &self,
+        req: &GetDatabaseRequest,
+    ) -> Result<Option<GetDatabaseResponse>, SysDbError> {
+        let mut stmt = Statement::new(
+            "SELECT ID, NAME, TENANT_ID FROM databases WHERE NAME = @name AND TENANT_ID = @tenant_id AND IS_DELETED = FALSE",
+        );
+        stmt.add_param("name", &req.name);
+        stmt.add_param("tenant_id", &req.tenant);
+
+        let mut tx = self
+            .client
+            .single()
+            .await
+            .map_err(|e| SysDbError::Spanner(e.to_string()))?;
+
+        let mut iter = tx
+            .query(stmt)
+            .await
+            .map_err(|e| SysDbError::Spanner(e.to_string()))?;
+
+        // Get the first row if it exists
+        if let Some(row) = iter
+            .next()
+            .await
+            .map_err(|e| SysDbError::Spanner(e.to_string()))?
+        {
+            let id: String = row
```

Comment:
I think it'd be nicer if we did something like this:
1) Have the Database and other domain structs implement the following trait
```
pub trait FromSpannerRow: Sized {
    fn from_row(row: &spanner::Row) -> Result<Self, SysDbError>;
}
```

and then over here we just call `Database::from_row(row)`.

We can go a step further and make a #derive() from this but perhaps that's an overkill.

Reply:
also don't need a new trait. The `TryFrom` trait already exists for this purpose